### ISSUE TITLE
Compilation fails after repeated use of `readRealParameter`: Too many open files

### ISSUE THREAD

**Issue by pstelzig** [OpenModelica/OpenModelica#10772]
### Description

After a number of simulation runs of models invoking the function `Modelica.Utilities.Examples.readRealParameter`, the compilation of the simulation binary fails with the message

```
[2] Not possible to open file "<path to parameter .txt file>" for reading: 
Too many open files

[3] 08:46:26 Übersetzung Fehler
[<Model name>]: Failed to evaluate function: Modelica.Utilities.Streams.readLine.
```

I am reporting this for a client and I have witnessed this behavior live on screen. However, I cannot reproduce the issue on my own computer. This behavior is probably not due to a faulty model implementation. The customer can reproduce the error by repeatedly executing simulation runs (click on "Simulate"-button) of `Modelica.Utilities.Examples.readRealParameterModel`. After 34 runs -- on compilation of the simulation binary of the 35th run -- the compilation fails with the above error message. **This problem does not appear with version 1.18**.

```
[2] 08:46:26 Fehler
Not possible to open file "C:\Users\[user]\AppData\Roaming\.openmodelica\libraries\Modelica 4.0.0+maint.om\Resources\Data\Utilities\Examples_readRealParameters.txt" for reading:
Too many open files

[3] 08:46:26 Übersetzung Fehler
[Modelica.Utilities.Examples: 325:8-325:61]: Failed to evaluate function: Modelica.Utilities.Streams.readLine.
```

After the issue occured, a restart of OMEdit is required to be able to re-simulate the very same model. On larger models of the customer, the issue occurs after only 5 simulation runs (i.e. clicks on "Simulate"), requiring restart of OMEdit.

**Observation** Interestinly the compilation (!) fails. I.e. something that is continuously running within OMEdit seems to accumulate file locks on the parameter file opened through `Modelica.Utilities.Examples.readRealParameter` or `Modelica.Utilities.Streams.readLine`. 

Although I am not familiar with how OMEdit and omc translate Modelica code into C, I doubt that invoking `Modelica.Utilities.Streams.close` changes the behavior (in fact, it seems it doesn't, see "Additional Context"). I think so because the error does not occur during runtime, and even if so: All simulation binaries are stopped after simulation runs, thus releasing all file locks they had. Does the mere translation into C and compilation really somehow execute `Modelica.Utilities.Streams.close` and could this release a file lock? 

### Steps to Reproduce

The behavior cannot be reproduced on arbitrary machines. It might thus be a local configuration issue, but I have no indication as to what might cause the issue. 

My client can reproduce the behavior by repeatedly executing simulation runs (click on "Simulate"-button) of `Modelica.Utilities.Examples.readRealParameterModel`. After 34 runs, on compilation of the simulation binary of the 35th run, the compilation fails with the above error message. 

### Expected Behavior

Models using `Modelica.Utilities.Examples.readRealParameter` can be compiled and simulated from within OMEdit arbitrarily many times in a row. 

### Version and OS

 - OpenModelica Version: 1.21
 - OS: Windows 10, 64 bit
 - Versions of used Modelica libraries if applicable: 4.0.0

### Additional Context
**I am aware of [issue 9968](https://github.com/OpenModelica/OpenModelica/issues/9968)**. However, even trying to improve `Modelica.Utilities.Examples.readRealParameter` with explicit calls on `Modelica.Utilities.Streams.close` did not change the behavior. 

Failed workaround: I modified `Modelica.Utilities.Examples.readRealParameter` with `Modelica.Utilities.Streams.close` as suggested and called `close` if the parameter is found or end of file is reached. 

**Directory structure of failed workaround**

```
realRealParameterFix
    |
    + params.txt
    |
    + readRealParameterFix.mo
    |
    + readRealParameterModelFix.mo
```

with files

**params.txt**

```
// Motor data

J        = 2.379   // inertia
w_rel0   = 1.5*2;  // relative angular velocity
phi_rel0 = pi/3;
```

**readRealParameterFix.mo**
```modelica
function readRealParameterFix "Read the value of a Real parameter from file"
  extends Modelica.Icons.Function;
  input String fileName "Name of file" annotation(Dialog(
                         loadSelector(filter="Text files (*.txt)",
                         caption="Open file in which Real parameters are present")));
  input String name "Name of parameter";
  output Real result "Actual value of parameter on file";

protected
  String line;
  String identifier;
  String delimiter;
  Integer nextIndex;
  Integer iline=1;
  Modelica.Utilities.Types.TokenValue token;
  String message = "in file \"" + fileName + "\" on line ";
  String message2;
  Boolean found = false;
  Boolean endOfFile=false;
algorithm
 (line, endOfFile) := Modelica.Utilities.Streams.readLine(fileName, iline);

  while not found and not endOfFile loop
    (token, nextIndex) := Modelica.Utilities.Strings.scanToken(line);
     if token.tokenType == Modelica.Utilities.Types.TokenType.NoToken then
        // skip line
        iline := iline + 1;
     elseif token.tokenType == Modelica.Utilities.Types.TokenType.IdentifierToken then
        if token.string == name then
           // name found, get value of "name = value;"
           message2 := message + String(iline);
           (delimiter,nextIndex) := Modelica.Utilities.Strings.scanDelimiter(line, nextIndex, {"="}, message2);
           (result,nextIndex)    := Modelica.Utilities.Examples.expression(line, nextIndex, message2);
           (delimiter,nextIndex) := Modelica.Utilities.Strings.scanDelimiter(line, nextIndex, {";", ""}, message2);
           Modelica.Utilities.Strings.scanNoToken(line, nextIndex, message2);
           found := true;
        else
           // wrong name, skip line
           iline := iline + 1;
        end if;
     else
        // wrong token
        Modelica.Utilities.Strings.syntaxError(line, nextIndex, "Expected identifier " + message + String(iline));
     end if;
     
     // read next line
     (line, endOfFile) := Modelica.Utilities.Streams.readLine(fileName, iline);
  end while;
  
  if not found then
     Modelica.Utilities.Streams.error("Parameter \"" + name + "\" not found in file \"" + fileName + "\"");
  end if;
  
  if found or endOfFile then
      Modelica.Utilities.Streams.close(fileName);
  end if;  
end readRealParameterFix;
```

**readRealParameterExampleFix.mo**
```modelica
model readRealParameterModelFix
  "Demonstrate usage of Examples.readRealParameter/.expression"

  extends Modelica.Icons.Example;

  parameter String file = Modelica.Utilities.Files.loadResource(
  "modelica://readRealParameterModelFix/params.txt")
    "File on which data is present"
        annotation(Dialog(loadSelector(filter="Text files (*.txt)",
                      caption="Open text file to read parameters of the form \"name = value\"")));
  parameter Modelica.Units.SI.Inertia J = readRealParameterFix(file, "J")
    "Inertia";
  parameter Modelica.Units.SI.Angle phi_rel0 = readRealParameterFix(file, "phi_rel0")
    "Relative angle";
  parameter Modelica.Units.SI.AngularVelocity w_rel0 = readRealParameterFix(file, "w_rel0")
    "Relative angular velocity";
equation
  when terminal() then
    Modelica.Utilities.Streams.close(file);
  end when;

  annotation (preferredView="text", Documentation(info="<html>
<p>
Model that shows the usage of Examples.readRealParameter and Examples.expression.
The model has 3 parameters and the values of these parameters are read
from a file.
</p>
</html>"), experiment(StopTime=1.01));
end readRealParameterModelFix;
```

**Comment by perost** [comment]
The compiler does indeed sometimes call external functions during compilation when necessary, for example to determine the size of an array variable or the value of a constant. The file I/O functions in the MSL usually caches opened files to avoid opening the same file over and over, but it seems that's not working properly when the compiler calls the external functions itself.

To call the external functions we load the shared library they're defined in and then use libffi to call the function, after which we unload the shared library again. I assume that's why the file caching doesn't work, since the cache is part of the ExternalModelicaC shared library. It also explains why explicitly closing the file doesn't work, since that also relies on the cache.

I made a MWE that demonstrates the issue (it needs a file `params.txt` in the same folder with at least one line in it, actual content doesn't matter):
```
model M
  constant String file = Modelica.Utilities.Files.loadResource("modelica://M/params.txt");
  constant String s[10000] = {Modelica.Utilities.Streams.readLine(file, i-i+1) for i in 1:size(s, 1)};
end M;
```
This model causes the compilation to fail after about 1000 calls to `readLine` for me, though I expect the limit will vary depending on OS and other factors. But curiously the error I get is:
```
Modelica 4.0.0+maint.om/Utilities/Streams.mo:68:3-96:15:writable] Error: External function ‘ModelicaInternal_readLine‘ could not be found in any of the given shared libraries:
--- list of various ModelicaExternalC.so files ---
```
But as mentioned this is after about 1000 calls to that function, so I wonder if the shared library is perhaps not unloaded properly and we hit some limit on number of shared libraries open? If that's the case then we should make sure it really is unloaded.

But besides that we probably shouldn't unload shared libraries during the compilation, but rather keep them open and only close them once the compilation is completed.

**Comment by sjoelund** [comment]
You should be able to list the open files through /proc/self/fd/ (or /proc/pid/fd/ for the process ID you are looking at). In order to open a shared library, it also needs to open a file so I guess the error makes sense.
It would also make sense to have an `la_objclose` in the ModelicaExternalC, but that would be an upstream change

**Comment by perost** [comment]
#10775 implements some simple caching of shared libraries so that we now only open a shared library once during compilation. That fixes my MWE at least, and seems to allow an arbitrary number of evaluations of external functions. It also fixes another subtle issue we had, where we didn't close a loaded library if it didn't contain the function we were looking for.

If nothing goes wrong it should be merged in a bit and be included in tomorrows nightly build if you want to try it out.

**Comment by pstelzig** [comment]
I have another recent update: Resetting the Options also changes the behavior. Why: I have no idea... Obviously it cannot not fix the issue that @perost could reproduce with the MWE and wrote the fix for, but it does something. 

Shortly after I submitted this issue here I poked through some other reports and found a post from @casella in [ticket 10591](https://github.com/OpenModelica/OpenModelica/issues/10591#issuecomment-1517792452)  and somewhat randomly suggested to my client that it might be worthwile resetting the options through the "Reset" button. 

With the resetted options: Instead of crashing on the 35th click on "Simulate" for `readRealParameterModel`, it now runs well over 50 times. Also, the client's larger proprietary models do not show the issue so far. So it seems that the old settings file carried over from 1.18 to 1.21 did something to the compilation process. 

But: Better solving the underlying real issue, so again thanks @perost! I'll try out the nightly build.
